### PR TITLE
Make pseudo fd for each file opening and each pseudo fd has own upload info

### PR DESCRIPTION
**Pull request by ggtakec**
## Relevant Issue (if applicable)
n/a

## Details
This PR is a fix for managing file descriptors that will be needed for future fixes in s3fs.
_I'm improving the performance of s3fs and want to update the correction part in small pieces, so first I will post the code without upload logic change._

### NOTE
This PR has a lot of modified code, but it is separated into the following two commits and put together as one PR.
The reason for posting them together is that if these are posted two separately, we may not understand the purpose of the PR.
Please do not squash when merging.

### Current code
When a file(S3 object) is opened, the file descriptor(issued by the system(os)) to the local file of the cache file(or temporary file) managed by s3fs is returned to FUSE.
s3fs always returns the same file descriptor, even if the file is opened at the same time.
(However, the number of open is kept as the number of references)
And s3fs cannot distinguish system calls(and callers) using fd for the same file.

### [Commit] Introduced pseudo fd and separated fd for each file opening
Changed s3fs to return a different pseudo file descriptor each time a file is opened at the same time.
This allows s3fs to recognize between callers, and system calls with fd can be processed depending on the caller.
Also, it has been fixed so that the file open flag(read/write, etc.) can be retained.

Note that there is only one cache file(and temporary file), even if the files are opened at the same time and different pseudo fd are issued.
In other words, pseuto fd is issued as a reference to one physical file.

### [Commit] Added info object about multipart uploading for each pseudo fd
The multi-part upload id and etag managed by the FdEntity object have been changed to be managed in units of pseudo fd.
I changed the information of multipart upload to be managed individually, but please note that there is only one cache file(or temporary file) for one file.

### CODE REVIEW

File: src/s3fs.cpp [hunk 1]
```diff
@@ -2316,13 +2316,10 @@ static int s3fs_read(const char* _path, char* buf, size_t size, off_t offset, st
 
     AutoFdEntity autoent;
     FdEntity*    ent;
-    if(NULL == (ent = autoent.ExistOpen(path, static_cast<int>(fi->fh)))){
+    if(NULL == (ent = autoent.GetExistFdEntiy(path, static_cast<int>(fi->fh)))){
```
[the remark this answers is not in the document]
Comment (ggtakec) [reply src/s3fs.cpp]: Thanks, it has been fixed.

**Comment by ggtakec** [comment]
@gaul 
The code before this PR returned the same file discriptor no matter how many times the file was opened.
And that fd was the system fd of the cache file(or temporary file).
It returns the same fd, but internally it was processing to increase or decrease the number of references.

This PR will now return the fd(not the system, pseudo) issued by s3fs each time the file is opened.
It also retains the file flags at the time of each opening.
Also, this pseudo fd is now treated as an object(PseudoFdInfo class).

The real purpose of this PR is to be able to manage the uploaded area  for each fd.
(But now, this PR is the same process as the code before PR in this PR code, but we plan to fix it in the future.)

**Comment by ggtakec** [comment]
The test is failing now, so please wait for a while as I will fix it.
